nalgebra-sparse/src/csr.rs

```diff
@@ -170,6 +170,24 @@ impl<T> CsrMatrix<T> {
         Self::try_from_pattern_and_values(pattern, values)
     }
 
+    /// Try to construct a CSR matrix from raw CSR data with unsorted columns.
+    pub fn try_from_unsorted_csr_data(
+        num_rows: usize,
+        num_cols: usize,
+        row_offsets: Vec<usize>,
+        col_indices: Vec<usize>,
+        values: Vec<T>,
+    ) -> Result<Self, SparseFormatError> {
+        let sorted_num_cols: Vec<usize> = row_offsets[0..row_offsets.len() - 1]
+            .iter()
+            .enumerate()
+            .flat_map(|(index, &offset)| {
+                Self::sorted(col_indices[offset..row_offsets[index + 1]].to_vec())
+            })
+            .collect();
```

Comment:
Here we need to keep in mind that the input we receive has *not* been validated. In particular, we *cannot* assume that `row_offsets.len() > 0` and that `col_index[offset..row_offsets[index + 1]]` won't panic. So we need to do the same input validation as `::try_from_csr_data`, with the exception of the requirement that column indices need to be sorted.

Moreover, we allocate a `Vec` in each `flat_map` invocation, and similarly the whole result is collected (so the whole thing is allocated again). This seems a little unnecessary: after all we don't need any more storage, we only need to sort subcollections in the input array.

So instead of using `.flat_map` and `.collect()` you could instead directly sort the corresponding entries in `col_indices`, i.e. something like
```rust
for (index, &offset) in ... {
   col_indices[offset .. row_offsets[index + 1]).sort_unstable();
}
```

Reply:
Thank you for feedback @Andlon!
I've changed my initial proposal according to your concerns. I hope it will bring the PR closer to a merge.
For example, I created a `first_and_last_offsets_are_ok` function to the base::helper which can be used in order to check whether the row_offsets is valid. But I was still not sure if it would be the best place for a new helper function.